<?xml version="1.0"?>

<robot name="al5d" xmlns:xacro="http://www.ros.org/wiki/xacro">
	
    <xacro:arg name="name" default="al5d"/>
 <xacro:property name="stance_height" value="1" />

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find al5d_simulation)/urdf/al5d.gazebo" />
  <!-- Import Rviz colors -->
  <xacro:include filename="$(find al5d_simulation)/urdf/materials.xacro" />


    <link name="world"/>
    <joint name="world_to_stance" type="fixed">
        <parent link="world"/>
        <child link="stance"/>
        <origin xyz="$(arg pose)" rpy="0 0 3.14159265359" />
    </joint>

    <link name="stance">
        <collision>  
            <origin xyz="0 0 ${stance_height/2}" rpy="0 0 0" />       
            -<geometry>                             
                <cylinder length="${stance_height}" radius="0.10" />            
            </geometry>   
        </collision>
        <visual>      
            <origin xyz="0 0 ${stance_height/2}" rpy="0 0 0" />       
            <geometry>  
                <cylinder length="${stance_height}" radius="0.10" />            
            </geometry>  
        </visual>
        <inertial>
            <origin xyz="0 0 ${stance_height/2}" rpy="0 0 0" /> 
            <mass value="1"/>
            <inertia
                ixx="1.0" ixy="0.0" ixz="0.0"
                iyy="1.0" iyz="0.0"
                izz="1.0"/>
        </inertial>
    </link>
    
    <joint name="stance_to_base" type="fixed">
        <parent link="stance"/>
        <child link="base_link"/>
        <origin xyz="0 0 ${stance_height}" rpy="0 0 0" />
        <axis xyz="0 0 0" />
    </joint>

	<link name="base_link">
		<collision>
      		<origin xyz="0 0 0.02" rpy="0 0 0" />
      		<geometry>
				<cylinder length="0.04" radius="0.095" />
     		 </geometry>
    	</collision>
		<visual>
			<geometry>
				<cylinder length="0.04" radius="0.095" />
			</geometry>
			<material name="black">
				<color rgba="0 0 0 1" />
			</material>
			<origin xyz="0 0 0.02" rpy="0 0 0" />
		</visual>
		<inertial>
      		<origin xyz="0 0 0.02" rpy="0 0 0" />
      		<mass value="1"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    		</inertial>
	</link>

	<link name="turret">
		<collision>
      		<origin xyz="0 0 0.0" rpy="0 0 0" />
      		<geometry>
				<cylinder length="0.005" radius="0.095" />
     		</geometry>
    	</collision>

		<visual>
			<origin xyz="0 0 0.0" rpy="0 0 0" />
			<geometry>
				<cylinder length="0.005" radius="0.095" />
			</geometry>
			<material name="black" />
		</visual>

		<inertial>
      		<origin xyz="0 0 0.0" rpy="0 0 0" />
      		<mass value="1"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>
	</link>

	<link name="upperarm">
		<visual>
			<geometry>			
				<box size="0.02 0.08 0.18" />
			</geometry>
			<material name="black" />
			<origin xyz="0 0 0.09" rpy="0 0 0" />
		</visual>	

		<collision>
      		<origin xyz="0 0 0.09" rpy="0 0 0" />
      		<geometry>
				<box size="0.02 0.08 0.18" />
     		 </geometry>
    	</collision>

		<inertial>
      		<origin xyz="0 0 0.09" rpy="0 0 0" />
      		<mass value="0.6"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>
	</link>

	<link name="forearm" >
		<visual>
			<geometry>
				<cylinder length="0.20" radius="0.015" />
			</geometry>
			<material name="silver" >
				<color rgba="0.9 0.9 0.9 1" />
			</material>
			<origin xyz="0 0 0.10" rpy="0 0 0" />
		</visual>

		<collision>
      		<origin xyz="0 0 0.10" rpy="0 0 0" />
      		<geometry>
				<cylinder length="0.20" radius="0.015" />
     		 </geometry>
    	</collision>	

    	<inertial>
      		<origin xyz="0 0 0.10" rpy="0 0 0" />
      		<mass value="0.4"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>	
	</link>

	<link name="wrist">
		<visual>
			<geometry>			
				<box size="0.02 0.05 0.06" />
			</geometry>
			<material name="black" />
			<origin xyz="0 0 0.03" rpy="0 0 0" />
		</visual>

		<collision>
      		<origin xyz="0 0 0.03" rpy="0 0 0" />
      		<geometry>
				<box size="0.02 0.05 0.06" />
     		 </geometry>
    	</collision>	

    	<inertial>
      		<origin xyz="0 0 0.03" rpy="0 0 0" />
      		<mass value="0.4"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>	
	</link>

	<link name="hand">
		<visual>
			<geometry>
				<box size="0.02 0.07 0.025" />
			</geometry>
			<material name="black" />
			<origin xyz="0 0 0.0125" rpy="0 0 0" />
		</visual>

		<collision>
      		<origin xyz="0 0 0.0125" rpy="0 0 0" />
      		<geometry>
				<box size="0.02 0.07 0.025" />
     		 </geometry>
    	</collision>	

    	<inertial>
      		<origin xyz="0 0 0.0125" rpy="0 0 0" />
      		<mass value="0.1"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>	
	</link>

	<link name="gripper_left">
		<visual>
			<geometry>
				<box size="0.02 0.02 0.03" />
			</geometry>
			<origin xyz="0 0 0.015" rpy="0 0 0" />
			<material name="gray" >
				<color rgba="0.3 0.3 0.3 1" />
			</material>
		</visual>

		<collision>
      		<origin xyz="0 0 0.015" rpy="0 0 0" />
      		<geometry>
				<box size="0.02 0.02 0.03" />
     		 </geometry>
    	</collision>	

    	<inertial>
      		<origin xyz="0 0 0.015" rpy="0 0 0" />
      		<mass value="0.1"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>	
	</link>

	<link name="gripper_right">
		<visual>
			<geometry>
				<box size="0.02 0.02 0.03" />
			</geometry>
			<origin xyz="0 0 0.015" rpy="0 0 0" />
			<material name="gray" />
		</visual>

		<collision>
      		<origin xyz="0 0 0.015" rpy="0 0 0" />
      		<geometry>
				<box size="0.02 0.02 0.03" />
     		 </geometry>
    	</collision>	

    	<inertial>
      		<origin xyz="0 0 0.015" rpy="0 0 0" />
      		<mass value="0.1"/>
      		<inertia
	  			ixx="1.0" ixy="0.0" ixz="0.0"
	  			iyy="1.0" iyz="0.0"
	  			izz="1.0"/>
    	</inertial>	

	</link>

	<joint name="base_link2turret" type="continuous">
		<parent link="base_link" />
		<child link="turret" />
		<origin xyz="0 0 0.045" rpy="0 0 ${pi/2}" />
		<axis xyz="0 0 1" />
		<limit effort="1000.0" lower="${-pi/2}" upper="${pi/2}" velocity="1.08"/>
	</joint>

	<joint name="turret2upperarm" type="continuous">
		<parent link="turret" />
		<child link="upperarm" />
		<origin xyz="0 0 0.02" rpy="0 0 0" />
		<axis xyz="0 -1 0" />
		<limit effort="1000.0" lower="${-pi/3}" upper="${(pi/2)/3}" velocity="0.84"/>
	</joint>

	<joint name="upperarm2forearm" type="continuous">
		<parent link="upperarm" />
		<child link="forearm" />
		<origin xyz="0 0 0.18" rpy="0 0 0" />
		<axis xyz="0 -1 0" />
		<limit effort="1000.0" lower="${-pi/2-pi/4}" upper="0" velocity="1.38"/>
	</joint>

	<joint name="forearm2wrist" type="continuous">
		<parent link="forearm" />
		<child link="wrist" />
		<origin xyz="0 0 0.20" rpy="0 0 0" />
		<axis xyz="0 -1 0" />
		<limit effort="1000.0" lower="${-pi/2}" upper="${pi/2}" velocity="1.2"/>
	</joint>

	<joint name="wrist2hand" type="continuous">
		<parent link="wrist" />
		<child link="hand" />
		<origin xyz="0 0 0.06" rpy="0 0 0" />
		<axis xyz="0 0 1" />
		<limit effort="1000.0" lower="${-pi/2}" upper="${pi/2}" velocity="1"/>
	</joint>

	<joint name="gripper_left2hand" type="prismatic">
		<parent link="hand" />
		<child link="gripper_left" />
		<origin xyz="0 0.025 0.025" rpy="0 0 0" />
		<limit effort="1000.0" lower="-0.02" upper="0.02" velocity="0.5" />
		<axis xyz="0 -1 0" />
	</joint>
	
	<joint name="gripper_right2hand" type="prismatic">
		<parent link="hand" />
		<child link="gripper_right" />
		<origin xyz="0 -0.025 0.025" rpy="0 0 0" />
		<limit effort="1000.0" lower="-0.02" upper="0.02" velocity="0.5" />
		<axis xyz="0 1 0" />
	</joint>

	<transmission name="base_link2turretTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="base_link2turret">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="base_link2turretServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="turret2upperarmTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="turret2upperarm">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="turret2upperarmServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="upperarm2forearmTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="upperarm2forearm">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="upperarm2forearmServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="forearm2wristTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="forearm2wrist">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="forearm2wristServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="wrist2handTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="wrist2hand">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="wrist2handServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="gripper_left2handTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="gripper_left2hand">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="gripper_left2handServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

	<transmission name="gripper_right2handTran">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="gripper_right2hand">
	    	<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="gripper_right2handServo">
       		<hardwareInterface>EffortJointInterface</hardwareInterface>
        	<mechanicalReduction>1</mechanicalReduction>
	    </actuator>
    </transmission>

</robot>
